r1cs-std/src/fields/fp3.rs

```diff
@@ -0,0 +1,932 @@
+                &b0_plus_2_b1_plus_4_b2,
+            )?
+        };
+
+        // v4 = a(∞)b(∞)   = a2 * b2
+        let v4 = self.c2.mul(cs.ns(|| "v2: a2 * b2"), &other.c2)?;
+
+        let two = P::Fp::one().double();
+        let six = two.double() + &two;
+        let mut two_and_six = [two, six];
+        algebra::fields::batch_inversion(&mut two_and_six);
+        let (two_inverse, six_inverse) = (two_and_six[0], two_and_six[1]);
+
+        let half_v0 = v0.mul_by_constant(cs.ns(|| "half_v0"), &two_inverse)?;
+        let half_v1 = v1.mul_by_constant(cs.ns(|| "half_v1"), &two_inverse)?;
+        let one_sixth_v2 = v2.mul_by_constant(cs.ns(|| "v2_by_six"), &six_inverse)?;
+        let one_sixth_v3 = v3.mul_by_constant(cs.ns(|| "v3_by_six"), &six_inverse)?;
+        let two_v4 = v4.double(cs.ns(|| "2 * v4"))?;
+
+        // c0 = v0 + β((1/2)v0 − (1/2)v1 − (1/6)v2 + (1/6)v3 − 2v4)
+        let c0 = {
+            let c0_cs = &mut cs.ns(|| "c0");
+
+            // No constraints, only get a linear combination back.
+            let temp = half_v0
+                .sub(c0_cs.ns(|| "sub1"), &half_v1)?
+                .sub(c0_cs.ns(|| "sub2"), &one_sixth_v2)?
+                .add(c0_cs.ns(|| "add3"), &one_sixth_v3)?
+                .sub(c0_cs.ns(|| "sub4"), &two_v4)?;
+            let non_residue_times_inner =
+                temp.mul_by_constant(&mut c0_cs.ns(|| "mul5"), &P::NONRESIDUE)?;
+            v0.add(c0_cs.ns(|| "add6"), &non_residue_times_inner)?
+        };
+
+        // −(1/2)v0 + v1 − (1/3)v2 − (1/6)v3 + 2v4 + βv4
+        let c1 = {
+            let c1_cs = &mut cs.ns(|| "c1");
+            let one_third_v2 = one_sixth_v2.double(&mut c1_cs.ns(|| "v2_by_3"))?;
+            let non_residue_v4 =
+                v4.mul_by_constant(&mut c1_cs.ns(|| "mul_by_beta"), &P::NONRESIDUE)?;
+
+            let result = half_v0
+                .negate(c1_cs.ns(|| "neg1"))?
+                .add(c1_cs.ns(|| "add2"), &v1)?
+                .sub(c1_cs.ns(|| "sub3"), &one_third_v2)?
+                .sub(c1_cs.ns(|| "sub4"), &one_sixth_v3)?
+                .add(c1_cs.ns(|| "sub5"), &two_v4)?
+                .add(c1_cs.ns(|| "sub6"), &non_residue_v4)?;
+            result
+        };
+
+        // -v0 + (1/2)v1 + (1/2)v2 −v4
+        let c2 = {
+            let c2_cs = &mut cs.ns(|| "c2");
+            let half_v2 = v2.mul_by_constant(&mut c2_cs.ns(|| "mul1"), &two_inverse)?;
+            let result = half_v1
+                .add(c2_cs.ns(|| "add1"), &half_v2)?
+                .sub(c2_cs.ns(|| "sub1"), &v4)?
+                .sub(c2_cs.ns(|| "sub2"), &v0)?;
+            result
+        };
+
+        Ok(Self::new(c0, c1, c2))
+    }
+
+    /// Use the Toom-Cook-3x method to compute multiplication.
+    #[inline]
+    fn square<CS: ConstraintSystem<ConstraintF>>(
+        &self,
+        mut cs: CS,
+    ) -> Result<Self, SynthesisError> {
+        // Uses Toom-Cool-3x multiplication from
```

Comment:
```suggestion
        // Uses Toom-Cook-3x multiplication from
```

Reply:
Ah, good catch. That typo is actually copied from fp6_3over2. I'll correct it there as well.